### PR TITLE
Add 3 new tutorials for TMVA deep learning:

### PR DESCRIPTION
**Pull request by lmoneta**
- TMVA_Higgs_Classification.C

   a tutorial using the public Higgs UCI dataset for a classification problem using a Deep Neural network from TMVA, that is  made with fully connected layers

 - TMVA_CNN_Classification.C

 Tutorial showing the usage of Convolutional neural network in TMVA.
 The macro generates on the fly some toys images (size 16x16) of two different classes and then a
 convolutional neural network is used for their classification.
 This example builds and uses also a CNN built on the fly using Keras through the ROOT PyKeras package
 This example shows also how to use a batch normalization layer in TMVA

- TMVA_RNN_Classification.C

 Tutorial showing the usage of Recurrent neural network in TMVA.
 Toys time dependednt data of two different classes are generated on the fly and then
 a recurrent neural network is used for classification.
 Both TMVA and PyKeras networks are built and used.
 The network uses by default one LSTM layer, but optionally it can be built with
 a simple RNN or a GRU layer or also 3 different recurrent networks for each recurrent layer type
 can be made

### CODE REVIEW

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+
+/// Here we book the new DNN of TMVA if we have support in ROOT. We will use GPU version if ROOT is enabled with GPU
+
+
+   /***
+
+## Booking Deep Neural Network
+
+Here we define the option string for builfing the Deep Neural network model.
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:181]:
```suggestion
Here we define the option string for building the Deep Neural network model.
```

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+#### 1. Define DNN layout
+
+The DNN configuration os defined using a string. Note that whitespaces between characters are not allowed.
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:185]:
```suggestion
The DNN configuration is defined using a string. Note that whitespaces between characters are not allowed.
```

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+We define the general DNN options concateneting in the final string the previously defined layout and training strategy.
+Note we use the ``":"`` separator to separate the different higher level options, as in the other TMVA methods.
+In addition to input layout, batch layout and training strategy we add now:
+
+- Type of Loss function (e.g. cross entropy)
+- Weight Initizalization (e.g XAVIER, XAVIERUNIFORM, NORMAL )
+- Variable Transformation
+- Type of Architecture (e.g. CPU, GPU, Standard)
+
+We can then book the DL method using the built otion string
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:234]:
```suggestion
We can then book the DL method using the built option string
```

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+ - Test Repetitions (the interval when the test error will be computed)
+
+
+#### 3. Define general DNN options
+
+We define the general DNN options concateneting in the final string the previously defined layout and training strategy.
+Note we use the ``":"`` separator to separate the different higher level options, as in the other TMVA methods.
+In addition to input layout, batch layout and training strategy we add now:
+
+- Type of Loss function (e.g. cross entropy)
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:229]:
```suggestion
- Type of Loss function (e.g. CROSSENTROPY)
```

Comment (AnirudhDagar) [reply tutorials/tmva/TMVA_Higgs_Classification.C:229]: Making comment on line 229 consistent with line 269 should be better imo.

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 5]
```diff
@@ -0,0 +1,467 @@
+
+/***
+
+    # TMVA Classification Example Using a Recurrent Neural Network
+
+    This is an example of using a RNN in TMVA. We do classification using a toy  data set
+    containing a time series of data sample ntimes and with dimension ndim
+   that is generated when running the provided function `MakeTimeData (nevents, ntime, ndim)
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:8]:
```suggestion
    This is an example of using a RNN in TMVA. We do classification using a toy data set
    containing a time series of data sample ntimes and with dimension ndim
    that is generated when running the provided function `MakeTimeData (nevents, ntime, ndim)
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+#include "TMVA/DataSetInfo.h"
+#include "TMVA/Config.h"
+#include "TMVA/MethodDL.h"
+
+
+#include "TFile.h"
+#include "TTree.h"
+
+// make some time data but not of fixed length.
+//  use a poisson with mu = 5 and troncated at 10
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:26]:
```suggestion
// use a poisson with mu = 5 and troncated at 10
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+   std::cout << "nthreads  = " << ROOT::GetThreadPoolSize() << std::endl;
+
+   TString inputFileName = "time_data_t10_d30.root";
+   // TString inputFileName = "/home/moneta/data/sample_images_32x32.gsoc.root";
+
+   bool fileExist = !gSystem->AccessPathName(inputFileName);
+
+   // if file does not exists create it
+   if (!fileExist) {
+      MakeTimeData(nTotEvts,ntime, ninput);
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:190]:
```suggestion
      MakeTimeData(nTotEvts, ntime, ninput);
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+
+         if (!use_rnn_type[i])
+            continue;
+
+         const char *rnn_type = rnn_types[i].c_str();
+
+         /// define the inputlayout string for RNN
+         /// the input data should be organize as   following:
+         //// input layout for RNN:    time x 1 x ndim
+         ///  batch layout for RNN     batchsize x time x ndim
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:289]:
```suggestion
         /// the input data should be organize as following:
         /// input layout for RNN:      time x 1 x ndim
         /// batch layout for RNN:     batchsize x time x ndim
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+
+         TString inputLayoutString = TString::Format("InputLayout=%d|1|%d", ntime, ninput);
+         TString batchLayoutString = TString::Format("BatchLayout=%d|%d|%d", batchSize, ntime, ninput);
+
+         /// Define RNN layer layout
+         ///  it should be   LayerType (RNN or LSTM or GRU) |  number of units | number of inputs | time steps | remember output (typically no=0 | return full sequence
+         TString rnnLayout = TString::Format("%s|10|%d|%d|0|1", rnn_type, ninput, ntime);
+
+         /// add after RNN a reshape layer (needed top flatten the output) and a dense layer with 64 units and a last one
+         /// NOte the last layer is linear because  when using Crossentropy a Sigmoid is applied
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:299]:
```suggestion
         /// Note the last layer is linear because when using Crossentropy a Sigmoid is applied already.
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+         TString trainingString1 = TString::Format("LearningRate=1e-3,Momentum=0.0,Repetitions=1,"
+                                             "ConvergenceSteps=5,BatchSize=%d,TestRepetitions=1,"
+                                             "WeightDecay=1e-2,Regularization=None,MaxEpochs=%d,"
+                                             "Optimizer=ADAM,DropConfig=0.0+0.+0.+0.",
+                                             batchSize,maxepochs);
+
+         TString trainingStrategyString("TrainingStrategy=");
+         trainingStrategyString += trainingString1; // + "|" + trainingString2
+
+         /// Define the full RN Noption string adding the final options for all network
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:312]:
```suggestion
         /// Define the full RNN option string adding the final options for all network
```

File: tutorials/tmva/TMVA_RNN_Classification.C [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+      for (int i = 0; i < 3; i++) {
+
+         if (use_rnn_type[i]) {
+
+            TString modelName = TString::Format("model_%s.h5", rnn_types[i].c_str());
+            TString trainedModelName = TString::Format("trained_model_%s.h5", rnn_types[i].c_str());
+
+            Info("TMVA_RNN_Classification", "Building recurrent keras model using a %s layer",rnn_types[i].c_str());
+            // create python script which can be executed
+            // crceate 2 conv2d layer + maxpool + dense
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_RNN_Classification.C:389]:
```suggestion
            // create 2 conv2d layer + maxpool + dense
```

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+//    loader->PrepareTrainingAndTestTree( mycut, "SplitMode=random:!V" );
+// To also specify the number of testing events, use:
+
+   loader->PrepareTrainingAndTestTree( mycuts, mycutb,
+                                       "nTrain_Signal=7000:nTrain_Background=7000:SplitMode=Random:NormMode=NumEvents:!V" );
+
+/***
+## Booking Methods
+
+Here we book the TMVA methods. We book firat a Likelihood based on KDE (Kernel Density Estimation), a Fischer discriminant, a BDT
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:138]:
```suggestion
Here we book the TMVA methods. We first book a Likelihood based on KDE (Kernel Density Estimation), a Fischer discriminant, a BDT
```

File: tutorials/tmva/TMVA_Higgs_Classification.C [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+void TMVA_Higgs_Classification() {
+
+   // options to control used methods
+
+   bool useLikelihood = true;    // likelihood based discriminant
+   bool useLikelihoodKDE = false;    // likelihood based discriminant
+   bool useFischer = true;       // Fischer discriminant
+   bool useMLP = false;          // Multi Layer Perceptron (old TMVA NN implementation)
+   bool useBDT = true;           // BOosted Decision Tree
+   bool useDL = true;            // TMVA Deep learning ( CPU ot GPU)
```
Comment (AnirudhDagar) [inline tutorials/tmva/TMVA_Higgs_Classification.C:36]:
```suggestion
   bool useDL = true;            // TMVA Deep learning ( CPU or GPU)
```

**Comment by lmoneta** [comment]
Thank you very much Anirudth for your review. I have included your suggested fixed in an update of the last commit
Cheers
